platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/ConfigurationCacheProblemsSummary.kt

```diff
@@ -52,23 +49,74 @@ enum class ProblemSeverity {
 
 
 internal
-class ConfigurationCacheProblemsSummary {
+class ConfigurationCacheProblemsSummary(
+
+    private
+    val maxCollectedProblems: Int = 4096
+
+) {
+    /**
+     * Reported more problems than can be collected.
+     */
+    private
+    var overflowed: Boolean = false
 
     private
-    val summary = AtomicReference(Summary.empty)
+    var problemCount: Int = 0
 
-    fun get(): Summary =
-        summary.get()
+    private
+    var failureCount: Int = 0
+
+    private
+    var suppressedCount: Int = 0
+
+    private
+    var uniqueProblems = ObjectOpenHashSet<UniquePropertyProblem>()
 
-    fun onProblem(problem: PropertyProblem, severity: ProblemSeverity): Boolean =
-        summary
-            .updateAndGet { it.insert(problem, severity) }
-            .overflowed.not()
+    private
+    var causes = ArrayList<Throwable>(maxCauses)
+
+    fun get(): Summary = synchronized(this) {
```

Comment:
💅 It doesn't look like this class needs the external synchronization protocol support, which `synchronized(this)` gives. Can we use a lock object for synchronization instead?

Reply:
Done in 2779442c0893f9087e564aec34f21e5f93827148.